### PR TITLE
feat: add formatNumber utility with K/M/B suffixes

### LINKED ISSUE #2: [Phase 1] Number formatting utility & shared helpers
## Summary
Create a shared `formatNumber` utility so all game values (Training Data, TD/s rates, costs) display consistently as 1K / 1M / 1B throughout the UI.

_Depends on #1_

## Acceptance criteria
- [ ] `src/utils/formatNumber.ts` exports a `formatNumber(n: number): string` function
- [ ] Values < 1,000 display as plain integers (e.g. `42`)
- [ ] Values >= 1,000 display with K suffix to 2 decimal places (e.g. `1.23K`)
- [ ] Values >= 1,000,000 display with M suffix (e.g. `4.56M`)
- [ ] Values >= 1,000,000,000 display with B suffix (e.g. `1.00B`)
- [ ] Vitest unit tests cover all boundary cases and edge cases (0, negatives, very large numbers)
- [ ] `npm run lint` passes with Biome

## Out of scope
- Localisation / locale-aware formatting
- Compact notation beyond B (T, Q, etc.) — can be added later

## Technical notes
- Keep the utility pure (no side effects) so it is easily testable
- Export from a barrel `src/utils/index.ts` so import paths stay clean
- CLAUDE.md: "Number formatting (1K, 1M, 1B) is needed early. Use a shared utility."

## Definition of Done
- Code implemented and tests passing
- PR opened and linked to this card
- Reviewer approved
- No regressions in related areas

### PR DESCRIPTION
**Pull request by 4sh-dev**
## Summary

Add a shared `formatNumber` utility so all game values (Training Data, TD/s rates, costs) display consistently with compact K / M / B suffixes throughout the UI.

## Changes

| File | Description |
|------|-------------|
| `src/utils/formatNumber.ts` | Pure function — formats numbers with K/M/B suffixes (2 decimal places) |
| `src/utils/index.ts` | Barrel export for clean import paths |
| `src/utils/formatNumber.test.ts` | 16 Vitest unit tests covering all tiers, boundaries, and edge cases |
| `vite.config.ts` | Configure Vitest (import from `vitest/config`) |
| `package.json` | Add Vitest 4.0.18, `test` and `test:watch` scripts |
| `.gitignore` | Ignore `*.tsbuildinfo` build artifacts |

### Formatting rules

| Range | Output | Example |
|-------|--------|---------|
| `n < 1,000` | Plain integer | `42` |
| `n >= 1,000` | 2dp + K | `1.23K` |
| `n >= 1,000,000` | 2dp + M | `4.56M` |
| `n >= 1,000,000,000` | 2dp + B | `1.00B` |

Negative values are supported (e.g. `-1.23K`). Sub-integer values are floored for the plain display tier.

## Testing

```bash
npm install
npm test        # 16 tests, all passing
npm run lint    # Biome — no errors
npm run build   # TypeScript + Vite — builds cleanly
```

Tests cover: zero, small integers, decimal truncation, exact tier boundaries (1K, 1M, 1B), mid-tier values, values just below boundaries, negative values at every tier, and the `-0` edge case.

## Story

Closes #2

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,9 @@
     "dev": "vite",
     "build": "tsc -b && vite build",
     "preview": "vite preview",
-    "lint": "biome check ."
+    "lint": "biome check .",
+    "test": "vitest run",
```
Comment (4sh-dev) [inline package.json]: **Suggestion: Enable tests in CI.** Now that Vitest is added, the placeholder in `.github/workflows/ci.yml` should be uncommented:

```yaml
      - run: npm test
```

Currently `ci.yml` has `# - run: npm test  # uncomment once a test runner is added`. Without this, tests won't run as a PR quality gate — a contributor could break tests and CI would still pass.

This can be done in this PR or as an immediate follow-up. Not blocking merge, but should land before Phase 2 PRs start opening.

File: vite.config.ts [hunk 2]
```diff
@@ -1,7 +1,11 @@
 import react from "@vitejs/plugin-react";
-import { defineConfig } from "vite";
+import { defineConfig } from "vitest/config";
 
 export default defineConfig({
   plugins: [react()],
   base: "/GLORP/",
+  test: {
+    globals: true,
```
Comment (4sh-dev) [inline vite.config.ts]: nit: `globals: true` injects `describe`, `it`, `expect` etc. as globals, but the test file explicitly imports them:

```ts
import { describe, expect, it } from "vitest";
```

The explicit imports are actually **better practice** (editor autocompletion, no ambient type magic). Consider removing `globals: true` to keep things consistent, or remove the imports from the test file if you prefer the globals approach.

Either way works — just pick one convention and stick with it.

File: src/utils/formatNumber.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * - Negative values are prefixed with "-" and formatted the same way.
+ */
+export function formatNumber(n: number): string {
+  if (n < 0) {
+    return `-${formatNumber(-n)}`;
+  }
+
+  for (const { threshold, suffix } of SUFFIXES) {
+    if (n >= threshold) {
+      return `${(n / threshold).toFixed(2)}${suffix}`;
```
Comment (4sh-dev) [inline src/utils/formatNumber.ts]: nit: `.toFixed(2)` can round values just below a tier boundary *up* past 1000, producing slightly awkward output:

```
formatNumber(999_999)     → "1000.00K"  (instead of "1.00M")
formatNumber(999_999_999) → "1000.00M"  (instead of "1.00B")
```

This happens because `999_999 / 1000 = 999.999` and `.toFixed(2)` rounds to `1000.00`, but the value is still below the 1M threshold.

The tests explicitly expect this behavior, so this is intentional — just flagging it as a cosmetic edge case. If it ever bothers users, a post-division check could bump to the next tier:

```ts
const scaled = n / threshold;
if (scaled >= 1000 && idx > 0) { /* use next tier */ }
```

No action needed now — the issue notes T/Q are out of scope, and this only affects a narrow band of values near each boundary.

**Review by 4sh-dev — COMMENTED** [review]
## Review Summary (from 4sh-reviewer / Remy)

> *Posting as COMMENT because GitHub blocks self-approval on cross-fork PRs from the same token.*

**Verdict: Approve** — all 7 acceptance criteria from issue #2 are satisfied. Code is clean, pure, and well-tested. Verified locally: 16 tests pass, lint clean, build succeeds.

---

### Acceptance Criteria — Issue #2

| # | Criterion | Status | Evidence |
|---|-----------|--------|----------|
| 1 | `src/utils/formatNumber.ts` exports `formatNumber(n: number): string` | **Pass** | Exported function with correct signature |
| 2 | Values < 1,000 as plain integers (e.g. `42`) | **Pass** | `Math.floor(n).toString()` — tested with 0, 1, 42, 999, decimals |
| 3 | Values >= 1,000 with K suffix, 2 decimals (e.g. `1.23K`) | **Pass** | `(n / 1000).toFixed(2) + "K"` — tested |
| 4 | Values >= 1,000,000 with M suffix (e.g. `4.56M`) | **Pass** | Tested with 1M, 4.56M |
| 5 | Values >= 1,000,000,000 with B suffix (e.g. `1.00B`) | **Pass** | Tested with 1B, 1.5B, 999B, 1T+ |
| 6 | Vitest tests cover boundaries and edge cases | **Pass** | 16 tests: boundaries, negatives, -0, very large numbers |
| 7 | `npm run lint` passes | **Pass** | Verified: 13 files checked, 0 issues |
| — | Barrel export from `src/utils/index.ts` | **Pass** | `export { formatNumber } from "./formatNumber"` |
| — | Pure function, no side effects | **Pass** | No imports, no mutation, recursive negative handling |

### Local Verification

```
✓ npm test     → 16 tests passed (137ms)
✓ npm run lint → 13 files checked, 0 issues
✓ npm run build → 749 modules, dist/ output
```

### What Looks Good

- **Clean implementation** — `SUFFIXES` array with descending thresholds is elegant and easy to extend later (T, Q, etc.)
- **`ReadonlyArray`** prevents accidental mutation of the suffix table
- **Negative handling via recursion** is elegant: `return \`-${formatNumber(-n)}\``
- **`-0` correctly returns `"0"`** — tested explicitly
- **Good test structure** — organized by tier (below 1K, K, M, B) plus negatives and edge cases
- **Vitest pinned at exact version** (`4.0.18`) — consistent with project conventions
- **`*.tsbuildinfo` added to `.gitignore`** — good housekeeping for `tsc -b`

### Issues

See 3 inline comments below:

1. **Enable `npm test` in CI** — the `ci.yml` placeholder should be uncommented now (most important)
2. **nit:** `globals: true` in vitest config is unused
3. **nit:** `.toFixed(2)` rounding near tier boundaries

None are blocking merge.

File: package.json (hunk 1 above)
Comment (4sh-dev) [reply package.json]: Good call — enabled `npm test` in CI in d9fb236. Tests now run as a PR quality gate.

**Review by 4sh-dev — COMMENTED** [review]
## Re-review (from 4sh-reviewer / Remy)

> *Reviewing updated commits `d9fb236` and `0716cf4`. Posting as COMMENT (self-approval limitation).*

**Verdict: READY TO MERGE** — all prior feedback addressed, all acceptance criteria still passing.

---

### Prior Feedback — All Resolved

| Issue | Status | Commit |
|-------|--------|--------|
| Enable `npm test` in CI | **Fixed** — placeholder uncommented, tests now run as a PR quality gate | `d9fb236` |
| Remove unused `globals: true` | **Fixed** — removed from `vite.config.ts`, explicit imports kept in test file | `0716cf4` |
| `.toFixed(2)` rounding near tier boundaries | **Acknowledged** — intentional, flagged for future improvement | n/a |

### Local Verification (on `0716cf4`)

```
✓ npm test     → 16 tests passed (129ms)
✓ npm run lint → 13 files checked, 0 issues
✓ npm run build → 749 modules, dist/ output
```

### CI Workflow

`ci.yml` now runs the full pipeline: `npm ci` → `npm run lint` → `npm run build` → `npm test`. Tests are a proper PR quality gate.

### Note on CI Status

GitHub shows CI status as "pending" with 0 checks — this is expected for cross-fork PRs (GitHub Actions doesn't automatically run `pull_request` workflows on PRs from forks for security reasons). The workflow is correctly configured and will run once the code lands on `main`. All checks verified locally.

---

**No remaining issues. This PR is ready to merge.**
